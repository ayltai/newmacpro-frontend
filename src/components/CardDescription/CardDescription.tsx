import { styled, } from '@mui/material';
import React, { FC, } from 'react';

import { default as Description, } from '../Description';
import type { CardDescriptionProps, } from './CardDescription.types';

const TruncatedDescription = styled(Description)({
    display         : '-webkit-box',
    overflow        : 'hidden',
    textOverflow    : 'ellipsis',
    WebkitBoxOrient : 'vertical',
    WebkitLineClamp : 2,
});

const CardDescription : FC<CardDescriptionProps> = ({
    children,
    ...rest
}) => (
    <TruncatedDescription {...rest}>
        {children}
    </TruncatedDescription>
);

export default CardDescription;
