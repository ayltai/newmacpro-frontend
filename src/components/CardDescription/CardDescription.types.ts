import { DescriptionProps, } from '../Description';

export interface CardDescriptionProps extends DescriptionProps {
}
