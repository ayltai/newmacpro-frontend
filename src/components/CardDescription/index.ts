export { default, } from './CardDescription';
export type { CardDescriptionProps, } from './CardDescription.types';
