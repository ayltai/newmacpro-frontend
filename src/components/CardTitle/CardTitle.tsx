import { styled, Typography, } from '@mui/material';
import React, { FC, } from 'react';

import type { CardTitleProps, } from './CardTitle.types';

const SingleLineTypography = styled(Typography)({
    display         : '-webkit-box',
    overflow        : 'hidden',
    textOverflow    : 'ellipsis',
    WebkitBoxOrient : 'vertical',
    WebkitLineClamp : 1,
});

const CardTitle : FC<CardTitleProps> = ({
    children,
    ...rest
}) => (
    <SingleLineTypography
        gutterBottom
        variant='h6'
        fontWeight='bold'
        {...rest}>
        {children}
    </SingleLineTypography>
);

export default CardTitle;
