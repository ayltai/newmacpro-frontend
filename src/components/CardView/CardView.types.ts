import { ReactElement, } from 'react';

import type { CardContentProps, } from '../CardContent';
import type { CardDescriptionProps, } from '../CardDescription';
import type { CardImageProps, } from '../CardImage';
import type { CardTitleProps, } from '../CardTitle';

export interface CardViewProps {
    align?           : 'left' | 'center' | 'right',
    children?        : ReactElement<CardContentProps | CardDescriptionProps | CardImageProps | CardTitleProps> | ReactElement<CardContentProps | CardDescriptionProps | CardImageProps | CardTitleProps>[],
    onClick?         : () => void,
    onMoreInfo?      : () => void,
    [ key : string ] : any,
}
