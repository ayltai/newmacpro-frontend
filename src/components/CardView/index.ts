export { default, } from './CardView';
export type { CardViewProps, } from './CardView.types';
