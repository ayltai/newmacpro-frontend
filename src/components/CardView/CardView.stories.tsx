import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';

import { POPULAR_APPS, } from '../../constants';
import CardContent from '../CardContent';
import CardDescription from '../CardDescription';
import CardImage from '../CardImage';
import CardTitle from '../CardTitle';
import { ListView, } from '../ListView';
import CardView from './CardView';

const meta : Meta<typeof CardView> = {
    component : CardView,
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        align    : {
            description : 'The alignment of the content.',
        },
        children : {
            control     : false,
            description : 'The content of the component.',
        },
        onClick  : {
            description : 'An event handler called when the component is clicked.',
        },
        onMoreInfo : {
            description : 'An event handler called when the "More Info" button is clicked.',
        },
    },
};

export default meta;

type Story = StoryObj<typeof CardView>;

/**
 * A CardView with default props.
 */
export const Default : Story = {
    args : {
        children : [
            <CardTitle key='title'>Firefox</CardTitle>,
            <CardDescription key='description'>A web browser</CardDescription>,
        ],
    },
};

/**
 * A CardView with an image and contents center aligned.
 */
export const WithImage : Story = {
    name : 'With image',
    args : {
        align    : 'center',
        children : [
            <CardTitle key='title'>Firefox</CardTitle>,
            <CardDescription key='description'>A web browser</CardDescription>,
            <CardImage
                key='image'
                height={180}
                imageUrl='/images/newmac-1.webp'
                altText='NewMac.Pro' />,
        ],
    },
};

/**
 * A CardView with contents left aligned.
 */
export const WithContent : Story = {
    name : 'With content',
    args : {
        align    : 'left',
        children : [
            <CardTitle key='title'>Firefox</CardTitle>,
            <CardDescription key='description'>A web browser</CardDescription>,
            <CardContent key='content'>
                <ListView
                    packages={POPULAR_APPS} />
            </CardContent>,
        ],
    },
};
