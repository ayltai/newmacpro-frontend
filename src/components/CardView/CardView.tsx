import { Box, Button, Card, CardActionArea, CardActions, CardContent, } from '@mui/material';
import React, { FC, ReactElement, } from 'react';
import { useTranslation, } from 'react-i18next';

import { default as Content, CardContentProps as ContentProps, } from '../CardContent';
import { default as CardDescription, CardDescriptionProps, } from '../CardDescription';
import { default as CardImage, CardImageProps, } from '../CardImage';
import { default as CardTitle, CardTitleProps, } from '../CardTitle';
import type { CardViewProps, } from './CardView.types';

/**
 * CardViews contain content and actions about a single subject.
 * @param align The alignment of the content.
 * @param children The content of the component.
 * @param onClick An event handler called when the component is clicked.
 * @param onMoreInfo An event handler called when the "More Info" button is clicked.
 * @param rest The other props passed down to the `CardView` component.
 */
const CardView : FC<CardViewProps> = ({
    align,
    children,
    onClick,
    onMoreInfo,
    ...rest
}) => {
    const { t, } = useTranslation();

    let title       : ReactElement<CardTitleProps>       | undefined;
    let description : ReactElement<CardDescriptionProps> | undefined;
    let image       : ReactElement<CardImageProps>       | undefined;

    const contents : ReactElement<ContentProps>[] = [];

    if (children) {
        if (Array.isArray(children)) {
            title       = children.find(child => child.type === CardTitle)       as ReactElement<CardTitleProps>;
            description = children.find(child => child.type === CardDescription) as ReactElement<CardDescriptionProps>;
            image       = children.find(child => child.type === CardImage)       as ReactElement<CardImageProps>;

            contents.push(...children.filter(child => child.type === Content));
        } else {
            if (children.type === CardTitle)       title       = children as ReactElement<CardTitleProps>;
            if (children.type === CardDescription) description = children as ReactElement<CardDescriptionProps>;
            if (children.type === CardImage)       image       = children as ReactElement<CardImageProps>;

            if (children.type === Content) contents.push(children as ReactElement<ContentProps>);
        }
    }

    return (
        <Card
            variant='elevation'
            {...rest}>
            <Box sx={{
                height        : '100%',
                display       : 'flex',
                flexDirection : 'column',
            }}>
                <Box sx={{
                    flex : 1,
                }}>
                    <CardActionArea
                        sx={{
                            height         : '100%',
                            display        : 'flex',
                            alignItems     : 'start',
                            justifyContent : align,
                        }}
                        onClick={onClick}>
                        <Box sx={{
                            width         : '100%',
                            display       : 'flex',
                            flexDirection : 'column',
                        }}>
                            {image}
                            <CardContent sx={{
                                width     : '100%',
                                textAlign : align,
                            }}>
                                {title}
                                {description}
                                <Box sx={{
                                    textAlign : 'left',
                                }}>
                                    {contents}
                                </Box>
                            </CardContent>
                        </Box>
                    </CardActionArea>
                </Box>
                <CardActions sx={{
                    display   : 'block',
                    textAlign : align,
                }}>
                    <Button
                        size='small'
                        variant='text'
                        color='info'
                        onClick={onMoreInfo}>
                        {t('package.more_info')}
                    </Button>
                </CardActions>
            </Box>
        </Card>
    );
};

export default CardView;
