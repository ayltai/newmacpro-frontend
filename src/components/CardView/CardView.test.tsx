import React from 'react';

import { fireEvent, render, } from '../../utils/test';
import CardDescription from '../CardDescription';
import CardTitle from '../CardTitle';
import CardView from './CardView';

const children = [
    <CardTitle key='title'>Firefox</CardTitle>,
    <CardDescription key='description'>A web browser</CardDescription>,
];

describe('<CardView />', () => {
    it('renders correctly', () => {
        expect(render(<CardView>{children}</CardView>)).toMatchSnapshot();
    });

    it('handles click event correctly', () => {
        const handleClick = jest.fn();

        const { getAllByRole, } = render(
            <CardView onClick={handleClick}>
                {children}
            </CardView>
        );

        fireEvent.click(getAllByRole('button')[0]);

        expect(handleClick).toBeCalledTimes(1);
    });

    it('handles more-info event correctly', () => {
        const handleMoreInfo = jest.fn();

        const { getAllByRole, } = render(
            <CardView onMoreInfo={handleMoreInfo}>
                {children}
            </CardView>
        );

        fireEvent.click(getAllByRole('button')[1]);

        expect(handleMoreInfo).toBeCalledTimes(1);
    });
});
