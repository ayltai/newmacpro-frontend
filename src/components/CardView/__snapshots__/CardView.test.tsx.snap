// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardView /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-wqbum7-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-1sm2s1z"
        >
          <div
            class="MuiBox-root css-1rr4qq7"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-1iwc3gj-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiBox-root css-4rmlxi"
              >
                <div
                  class="MuiCardContent-root css-1qf7ogf-MuiCardContent-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1ywgf30-MuiTypography-root"
                  >
                    Firefox
                  </h6>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1qnobji-MuiTypography-root"
                  >
                    A web browser
                  </p>
                  <div
                    class="MuiBox-root css-13brihr"
                  />
                </div>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1a42o16-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall css-1ujgr8z-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              package.more_info
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-wqbum7-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiBox-root css-1sm2s1z"
      >
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root css-1iwc3gj-MuiButtonBase-root-MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root css-4rmlxi"
            >
              <div
                class="MuiCardContent-root css-1qf7ogf-MuiCardContent-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1ywgf30-MuiTypography-root"
                >
                  Firefox
                </h6>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1qnobji-MuiTypography-root"
                >
                  A web browser
                </p>
                <div
                  class="MuiBox-root css-13brihr"
                />
              </div>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1a42o16-MuiCardActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeSmall MuiButton-textSizeSmall css-1ujgr8z-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            package.more_info
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
